### DISCUSSION TITLE
How to use the GraphQL API to find non-materialized partitions for an asset

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#18344] [Q&A] [answered]
is there a way using the GraphQL to find unmaterialized partitions? I have the start of my query in thread, but I'm not sure how to go about finding unmaterialized or failed partitions for an asset. We're on dagster version `1.4.3`

```
query PartitionStats {
  assetNodeOrError(assetKey: {path:"stripe_balance_transactions"}){
    ... on AssetNode {
      assetKey{
        path
      }
      assetPartitionStatuses {
      	__typename
      }
    }
  }
}
```

the asset in particular returns `TimePartitionStatuses`

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C01U954MEER/p1701202991280699?thread_ts=1701202991.280699&cid=C01U954MEER).*

**Comment by steffyd** [comment]
There's some additional context that I've linked in this bug here: https://github.com/dagster-io/dagster/issues/18369

**Comment by clairelin135** [comment, marked as answer]
@steffyd Responded to you in the other thread, but unfortunately there isn't a great way to do this via graphQL. GraphQL is used to support the UI, and queries are subject to change.

You can do this via the instance though:
```python
instance = DagsterInstance.get()
instance.get_status_by_partition(asset_key, partition_keys, partitions_def)
```

**Comment by danielgafni** [comment]
I want to programmatically create `DynamicPartitions` from production in a Branch Deployment (I want the partition keys to be copied). 

Is there currently any way of doing this?

**Comment by sergii-imu** [comment]
It is possible to get all partition keys for unmaterialized partitions like:

```graphql
query my_query {
  assetOrError(assetKey: { path: ["my_asset_key"] }) {
    __typename 
    ... on Asset {
      id
      key {path}
      definition {
        assetPartitionStatuses {
          __typename
          ... on DefaultPartitionStatuses {
            unmaterializedPartitions
          }
        }
      }
    }
    ... on AssetNotFoundError {
      message
    }
  }
}
```
